File: libraries/fary/ary1_dbnd.f
      SUBROUTINE ARY1_DBND( IDCB, STATUS )
*+
*  Name:
*     ARY1_DBND

*  Purpose:
*     Ensure that bounds (and dimensionality) information is available
*     for a data object.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_DBND( IDCB, STATUS )

*  Description:
*     The routine ensures that bounds and dimensionality information is
*     available for an entry in the DCB. The routine does nothing if
*     this information is already available. Otherwise, it obtains the
*     information by inspecting the data object itself and stores the
*     results in the DCB. Only those checks necessary to obtain the
*     bounds information are performed on the data object.

*  Arguments:
*     IDCB = INTEGER (Given)
*        Index of the data object entry in the DCB.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Do nothing if the information is already present in the DCB.
*     -  Ensure that form information is available for the data object.
*     -  Handle each form of array separately.
*     -  For primitive arrays, ensure that data type information is
*     available, then obtain the shape of the array.
*     -  Check that it is not scalar, and report an error if it is.
*     -  Enter the dimensionality and bounds information into the DCB,
*     padding the bounds with 1's if necessary.
*     -  For simple arrays, ensure that data type information and
*     component locators are available.
*     -  Obtain the shape of the non-imaginary component and check it is
*     not scalar. Report an error if it is.
*     -  If the array is complex, then obtain the shape of the imaginary
*     component and check it is the same as the non-imaginary component.
*     Report an error if it is not.
*     -  If there is no ORIGIN component, then supply default values for
*     it.
*     -  If there is an ORIGIN component, then obtain its type and shape
*     and check that these are valid, reporting an error if they are
*     not. Obtain the ORIGIN values.
*     -  Store the array bounds information in the DCB and note this
*     fact.

*  Copyright:
*     Copyright (C) 1989, 1990 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*     
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*     
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S Berry (JAC)
*     {enter_new_authors_here}

*  History:
*     12-JUN-1989 (RFWS):
*        Original version.
*     16-AUG-1989 (RFWS):
*        Changed initialisation of locators to use global constant.
*     23-AUG-1989 (RFWS):
*        Added missing space to error message.
*     12-FEB-1990 (RFWS):
*        Added support for primitive arrays.
*     26-APR-2006 (DSB):
*        Added support for scaled arrays.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Global Variables:
      INCLUDE 'ARY_DCB'          ! ARY_ Data Control Block
*        DCB_CPX( ARY__MXDCB ) = LOGICAL (Read)
*           Whether the array is complex.
*        DCB_DLOC( ARY__MXDCB ) = CHARACTER * ( DAT__SZLOC ) (Read)
*           Locator to non-imaginary data component.
*        DCB_FRM( ARY__MXDCB ) = CHARACTER * ( ARY__SZFRM ) (Read)
*           Form of the array.
*        DCB_ILOC( ARY__MXDCB ) = CHARACTER * ( DAT__SZLOC ) (Read)
*           Locator to imaginary data component.
*        DCB_KBND( ARY__MXDCB ) = LOGICAL (Read and Write)
*           Whether the DCB bounds information is up to date.
*        DCB_LBND( ARY__MXDIM, ARY__MXDCB ) = INTEGER (Write)
*           Lower bounds of array.
*        DCB_LOC( ARY__MXDCB ) = CHARACTER * ( DAT__SZLOC ) (Read)
*           Locator to data object.
*        DCB_NDIM( ARY__MXDCB ) = INTEGER (Write)
*           Number of array dimensions.
*        DCB_UBND( ARY__MXDIM, ARY__MXDCB ) = INTEGER (Write)
*           Upper bounds of array.

*  Arguments Given:
      INTEGER IDCB

*  Status:
      INTEGER STATUS             ! Global status

*  Local variables:
      CHARACTER * ( DAT__SZLOC ) LOCOR ! Locator to ORIGIN component
      CHARACTER * ( DAT__SZTYP ) TYPE ! HDS data type
      INTEGER DIMD( ARY__MXDIM ) ! Dimensions of DATA component
      INTEGER DIMI( ARY__MXDIM ) ! Dimensions of IMAGINARY_DATA compnt
      INTEGER DIMOR( 1 )         ! Dimensions of ORIGIN component
      INTEGER I                  ! Loop counter for dimensions
      INTEGER NDIMD              ! No. of DATA dimensions
      INTEGER NDIMI              ! No. of IMAGINARY_DATA dimensions
      INTEGER NDIMOR             ! Number of ORIGIN dimensions
      INTEGER ORIG( ARY__MXDIM ) ! Origin values
      LOGICAL THERE              ! Whether a component exists

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Do nothing if bounds information is ready available in the DCB.
      IF ( .NOT. DCB_KBND( IDCB ) ) THEN

*  Ensure that form information is available.
         CALL ARY1_DFRM( IDCB, STATUS )
         IF ( STATUS .NE. SAI__OK ) GO TO 9999

*  Primitive arrays.
*  ================
         IF ( DCB_FRM( IDCB ) .EQ. 'PRIMITIVE' ) THEN

*  Ensure that data type information is available, then obtain the
*  shape of the array.
            CALL ARY1_DTYP( IDCB, STATUS )
            CALL DAT_SHAPE( DCB_DLOC( IDCB ), ARY__MXDIM, DIMD, NDIMD,
     :                      STATUS )
            IF ( STATUS .NE. SAI__OK ) GO TO 9999

*  Check that it is not scalar, and report an error if it is.
            IF ( NDIMD .EQ. 0 ) THEN
               STATUS = ARY__NDMIN
               CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
               CALL ERR_REP( 'ARY1_DBND_PSCLR',
     :         'The array ^ARRAY is a scalar; it should have at ' //
     :         'least one dimension.', STATUS )
               GO TO 9999
            END IF

*  Enter the dimensionality and bounds information into the DCB, padding
*  the bounds with 1's if necessary.
            DCB_NDIM( IDCB ) = NDIMD
            DO 1 I = 1, NDIMD
               DCB_LBND( I , IDCB ) = 1
               DCB_UBND( I , IDCB ) = DIMD( I )
1           CONTINUE
            DO 2 I = NDIMD + 1, ARY__MXDIM
               DCB_LBND( I, IDCB ) = 1
               DCB_UBND( I, IDCB ) = 1
2           CONTINUE

*  Simple and scaled arrays.
*  =========================
         ELSE IF ( DCB_FRM( IDCB ) .EQ. 'SIMPLE' .OR.
     :             DCB_FRM( IDCB ) .EQ. 'SCALED' ) THEN

*  Ensure that data type information is available, then obtain the shape
*  of the non-imaginary array component.
            CALL ARY1_DTYP( IDCB, STATUS )
            CALL DAT_SHAPE( DCB_DLOC( IDCB ), ARY__MXDIM, DIMD, NDIMD,
     :      STATUS )
            IF ( STATUS .NE. SAI__OK ) GO TO 9999

*  Check that it is not scalar, and report an error if it is.
            IF ( NDIMD .EQ. 0 ) THEN
               STATUS = ARY__NDMIN
               CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
               CALL ERR_REP( 'ARY1_DBND_SCLR',
     :         'The DATA component in the array structure ^ARRAY ' //
     :         'is a scalar; it should have at least one dimension.',
     :         STATUS )
               GO TO 9999
            END IF

*  If the array is complex, then obtain the shape of the complex
*  component.
            IF ( DCB_CPX( IDCB ) ) THEN
               CALL DAT_SHAPE( DCB_ILOC( IDCB ), ARY__MXDIM, DIMI,
     :         NDIMI, STATUS )
               IF ( STATUS .NE. SAI__OK ) GO TO 9999

*  Report an error if the two components have different
*  dimensionalities.
               IF ( NDIMI .NE. NDIMD ) THEN
                  STATUS = ARY__NDMIN
                  CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
                  CALL MSG_SETI( 'NDIMD', NDIMD )
                  CALL MSG_SETI( 'NDIMI', NDIMI )
                  CALL ERR_REP( 'ARY1_DBND_NDIM',
     :            'The IMAGINARY_DATA component in the array ' //
     :            'structure ^ARRAY has a different number of ' //
     :            'dimensions (^NDIMI) to the corresponding DATA ' //
     :            'component (^NDIMD).', STATUS )
                  GO TO 9999

*  Report an error if any of the dimension sizes is different in the
*  two components.
               ELSE
                  DO 3 I = 1, NDIMD
                     IF ( DIMI( I ) .NE. DIMD( I ) ) THEN
                        STATUS = ARY__DIMIN
                        CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
                        CALL MSG_SETI( 'I', I )
                        CALL MSG_SETI( 'DIMD', DIMD( I ) )
                        CALL MSG_SETI( 'DIMI', DIMI( I ) )
                        CALL ERR_REP( 'ARY1_DBND_DIM',
     :                  'Dimension no. ^I of the IMAGINARY_DATA ' //
     :                  'component in the array structure ^ARRAY ' //
     :                  'has a different size (^DIMI) to the ' //
     :                  'corresponding dimension of the array''s ' //
     :                  'DATA component (^DIMD).', STATUS )
                        GO TO 9999
                     END IF
3                 CONTINUE
               END IF
            END IF

*  See if there is an ORIGIN component present.
            CALL DAT_THERE( DCB_LOC( IDCB ), 'ORIGIN', THERE, STATUS )
            IF ( STATUS .NE. SAI__OK ) GO TO 9999

*  If not, then supply default origin values.
            IF ( .NOT. THERE ) THEN
               DO 4 I = 1, NDIMD
                  ORIG( I ) = 1
4              CONTINUE

*  If there is, then obtain its type and shape.
            ELSE
               LOCOR = ARY__NOLOC
               CALL DAT_FIND( DCB_LOC( IDCB ), 'ORIGIN', LOCOR, STATUS )
               CALL DAT_TYPE( LOCOR, TYPE, STATUS )
               CALL DAT_SHAPE( LOCOR, 1, DIMOR, NDIMOR, STATUS )
               IF ( STATUS .EQ. SAI__OK ) THEN

*  Report an error if the ORIGIN component is not of type '_INTEGER'.
                  IF ( TYPE .NE. '_INTEGER' ) THEN
                     STATUS = ARY__TYPIN
                     CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
                     CALL MSG_SETC( 'BADTYPE', TYPE )
                     CALL ERR_REP( 'ARY1_DBND_OTYP',
     :               'The ORIGIN component in the array structure ' //
     :               '^ARRAY has an invalid HDS type of ' //
     :               '''^BADTYPE''; its type should be ''_INTEGER''.',
     :               STATUS )

*  Report an error if it is not 1-dimensional.
                  ELSE IF ( NDIMOR .NE. 1 ) THEN
                     STATUS = ARY__NDMIN
                     CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
                     CALL MSG_SETI( 'NDIM', NDIMOR )
                     CALL ERR_REP( 'ARY1_DBND_ONDI',
     :               'The ORIGIN component in the array structure ' //
     :               '^ARRAY has an invalid number of dimensions ' //
     :               '(^NDIM); this component should be 1-dimensional.',
     :               STATUS )

*  Report an error if it has the wrong number of elements.
                  ELSE IF ( DIMOR( 1 ) .NE. NDIMD ) THEN
                     STATUS = ARY__DIMIN
                     CALL DAT_MSG( 'ARRAY', DCB_LOC( IDCB ) )
                     CALL MSG_SETI( 'DIM', DIMOR( 1 ) )
                     CALL MSG_SETI( 'NDIMD', NDIMD )
                     CALL ERR_REP( 'ARY1_DBND_ODIM',
     :               'The ORIGIN component in the array structure ' //
     :               '^ARRAY has an invalid number of elements ' //
     :               '(^DIM); this number should match the ' //
     :               'dimensionality of the array''s DATA component ' //
     :               '(^NDIMD).', STATUS )
                  END IF

*  Obtain the ORIGIN values.
                  CALL DAT_GET1I( LOCOR, ARY__MXDIM, ORIG, DIMOR( 1 ),
     :            STATUS )
               END IF

*  Annul the locator to the ORIGIN component.
               CALL DAT_ANNUL( LOCOR, STATUS )
               LOCOR = ARY__NOLOC
            END IF
            IF ( STATUS .NE. SAI__OK ) GO TO 9999

*  Enter the dimensionality and bounds information into the DCB, padding
*  the bounds with 1's if necessary.
            DCB_NDIM( IDCB ) = NDIMD
            DO 5 I = 1, NDIMD
               DCB_LBND( I , IDCB ) = ORIG( I )
               DCB_UBND( I , IDCB ) = ORIG( I ) + DIMD( I ) - 1
5           CONTINUE
            DO 6 I = NDIMD + 1, ARY__MXDIM
               DCB_LBND( I, IDCB ) = 1
               DCB_UBND( I, IDCB ) = 1
6           CONTINUE

*  If the form information in the DCB was not valid, then report an
*  error.
         ELSE
            STATUS = ARY__FATIN
            CALL MSG_SETC( 'BADFORM', DCB_FRM( IDCB ) )
            CALL ERR_REP( 'ARY1_DBND_FORM',
     :      'Unsupported array form ''^BADFORM'' found in Data ' //
     :      'Control Block (internal programming error).', STATUS )
         END IF

*  Note if the bounds information is now known.
9999     CONTINUE
         DCB_KBND( IDCB ) = STATUS .EQ. SAI__OK
      END IF
       
*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_DBND', STATUS )

      END

File: libraries/cary/ary1Dbnd.c
#include <string.h>
#include "sae_par.h"
#include "ary1.h"
#include "star/hds.h"
#include "mers.h"
#include "ary_err.h"

void ary1Dbnd( AryDCB *dcb, int *status ) {
/*
*+
*  Name:
*     ary1Dbnd

*  Purpose:
*     Ensure that bounds (and dimensionality) information is available
*     for a data object.

*  Synopsis:
*     void ary1Dbnd( AryDCB *dcb, int *status )

*  Description:
*     The routine ensures that bounds and dimensionality information is
*     available for an entry in the DCB. The routine does nothing if
*     this information is already available. Otherwise, it obtains the
*     information by inspecting the data object itself and stores the
*     results in the DCB. Only those checks necessary to obtain the
*     bounds information are performed on the data object.

*  Parameters:
*     dcb
*        Pointer to the DCB.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.
*     5-FEB-2018 (DSB):
*        Use _INT64 in place of HDS_DIM_TYPE when checking that the ORIGIN
*        data type is legal. This is because people with older starlink
*        installations may still have HDS_DIM_TYPE set to _INTEGER, and
*        we want such people to be able to read new NDFs (relying on the
*        automatic type conversion provided by HDS to convert _INT64 to
*        _INTEGER when the ORIGIN values are accessed).

*-
*/

/* Local variables: */
   HDSLoc *locor = NULL;      /* Locator to ORIGIN component */
   char type[DAT__SZTYP+1];   /* HDS data type */
   hdsdim dimd[ARY__MXDIM];   /* Dimensions of DATA component */
   hdsdim dimi[ARY__MXDIM];   /* Dimensions of IMAGINARY_DATA compnt */
   hdsdim dimor;              /* Dimensions of ORIGIN component */
   hdsdim orig[ARY__MXDIM];   /* Origin values */
   int i;                     /* Loop counter for dimensions */
   int ndimd;                 /* No. of DATA dimensions */
   int ndimi;                 /* No. of IMAGINARY_DATA dimensions */
   int ndimor;                /* Number of ORIGIN dimensions */
   int there;                 /* Whether a component exists */
   size_t nel;                /* Number of transferred elements */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Do nothing if bounds information is ready available in the DCB. */
   if( !dcb->kbnd ){

/* Ensure that form information is available. */
      ary1Dfrm( dcb, status );
      if( *status != SAI__OK ) goto L999;

/* Primitive arrays.
   ================ */
      if( !strcmp( dcb->form, "PRIMITIVE" ) ){

/* Ensure that data type information is available, then obtain the shape of
   the array. */
         ary1Dtyp( dcb, status );
         datShape( dcb->dloc, ARY__MXDIM, dimd, &ndimd, status );
         if( *status != SAI__OK ) goto L999;

/* Check that it is not scalar, and report an error if it is. */
         if( ndimd == 0 ){
            *status = ARY__NDMIN;
            datMsg( "ARRAY", dcb->loc );
            errRep( "ARY1_DBND_PSCLR",
                    "The array ^ARRAY is a scalar; it should have at least"
                    "one dimension.", status );
            goto L999;
         }

/* Enter the dimensionality and bounds information into the DCB, padding
   the bounds with 1's if necessary. */
         dcb->ndim = ndimd;
         for( i = 0 ; i < ndimd; i++ ){
            dcb->lbnd[ i ] = 1;
            dcb->ubnd[ i ] = dimd[ i ];
         }
         for( ; i < ARY__MXDIM; i++ ){
            dcb->lbnd[ i ] = 1;
            dcb->ubnd[ i ] = 1;
         }

/* Simple, scaled and delta arrays.
   ================================ */
      } else if( !strcmp( dcb->form, "SIMPLE" ) ||
                 !strcmp( dcb->form, "SCALED" ) ||
                 !strcmp( dcb->form, "DELTA" ) ){

/* Ensure that data type information is available, then for non-DELTA
   arrays, obtain the shape of the non-imaginary array component. */
         ary1Dtyp( dcb, status );
         if( strcmp( dcb->form, "DELTA" ) ){
            datShape( dcb->dloc, ARY__MXDIM, dimd, &ndimd, status );

/* For delta arrays, the dimensions are defined by the ZAXIS, ZDIM, and
   FIRST_DATA components in the data object. */
         } else {
            ary1Dlshp( dcb->loc, ARY__MXDIM, dimd, &ndimd, status );
         }

         if( *status != SAI__OK ) goto L999;

/* Check that it is not scalar, and report an error if it is. */
         if( ndimd == 0 ){
            *status = ARY__NDMIN;
            datMsg( "ARRAY", dcb->loc );
            errRep( "ARY1_DBND_SCLR",
                    "The DATA component in the array structure ^ARRAY is a"
                    "scalar; it should have at least one dimension.",
                    status );
            goto L999;
         }

/* If the array is complex, then obtain the shape of the complex component. */
         if( dcb->complex ){
            datShape( dcb->iloc, ARY__MXDIM, dimi, &ndimi, status );
            if( *status != SAI__OK ) goto L999;

/* Report an error if the two components have different dimensionalities. */
            if( ndimi != ndimd ){
               *status = ARY__NDMIN;
               datMsg( "ARRAY", dcb->loc );
               msgSeti( "NDIMD", ndimd );
               msgSeti( "NDIMI", ndimi );
               errRep( "ARY1_DBND_NDIM",
                       "The IMAGINARY_DATA component in the array structure"
                       "^ARRAY has a different number of dimensions"
                       "(^NDIMI) to the corresponding DATA component"
                       "(^NDIMD).", status );
               goto L999;

/* Report an error if any of the dimension sizes is different in the two
   components. */
            } else {
               for( i = 0; i < ndimd; i++ ){
                  if( dimi[ i ] != dimd[ i ] ){
                     *status = ARY__DIMIN;
                     datMsg( "ARRAY", dcb->loc );
                     msgSeti( "I", i );
                     msgSeti( "DIMD", dimd[ i ] );
                     msgSeti( "DIMI", dimi[ i ] );
                     errRep( "ARY1_DBND_DIM",
                             "Dimension no. ^I of the IMAGINARY_DATA"
                             "component in the array structure ^ARRAY has a"
                             "different size (^DIMI) to the corresponding"
                             "dimension of the array's DATA component"
                             "(^DIMD).", status );
                     goto L999;
                  }
               }
            }
         }

/* See if there is an ORIGIN component present. */
         datThere( dcb->loc, "ORIGIN", &there, status );
         if( *status != SAI__OK ) goto L999;

/* If not, then supply default origin values. */
         if( !there ){
            for( i = 0; i < ndimd; i++ ){
               orig[ i ] = 1;
            }

/* If there is, then obtain its type and shape. */
         } else {
            locor = NULL;
            datFind( dcb->loc, "ORIGIN", &locor, status );
            datType( locor, type, status );
            datShape( locor, 1, &dimor, &ndimor, status );
            if( *status == SAI__OK ){

/* Report an error if the ORIGIN component is not of the correct type.
   We need to read HDS v4 and v5 files, so allow it to be _INT64 or
   _INTEGER. Note, use "_INT64" rather than "HDS_DIM_TYPE" because people
   with older starlink installations may still have HDS_DIM_TYPE set to
   _INTEGER, and we want such people to be able to read new NDFs. */
               if( strcmp( type, "_INT64" ) &&
                   strcmp( type, "_INTEGER" ) ){
                  *status = ARY__TYPIN;
                  datMsg( "ARRAY", dcb->loc );
                  msgSetc( "BADTYPE", type );
                  errRep( "ARY1_DBND_OTYP",
                          "The ORIGIN component in the array structure"
                          "^ARRAY has an invalid HDS type of '^BADTYPE';"
                          "its type should be '_INTEGER' or '_INT64'.",
                          status );

/* Report an error if it is not 1-dimensional. */
               } else if( ndimor != 1 ){
                  *status = ARY__NDMIN;
                  datMsg( "ARRAY", dcb->loc );
                  msgSeti( "NDIM", ndimor );
                  errRep( "ARY1_DBND_ONDI",
                          "The ORIGIN component in the array structure"
                          "^ARRAY has an invalid number of dimensions"
                          "(^NDIM); this component should be 1-dimensional.",
                          status );

/* Report an error if it has the wrong number of elements. */
               } else if( dimor != ndimd ){
                  *status = ARY__DIMIN;
                  datMsg( "ARRAY", dcb->loc );
                  msgSeti( "DIM", dimor );
                  msgSeti( "NDIMD", ndimd );
                  errRep( "ARY1_DBND_ODIM",
                          "The ORIGIN component in the array structure"
                          "^ARRAY has an invalid number of elements (^DIM);"
                          "this number should match the dimensionality of"
                          "the array's DATA component (^NDIMD).", status );
               }

/* Obtain the ORIGIN values. */
               HDSDIM_CODE(datGet1)( locor, ARY__MXDIM, orig, &nel, status );
            }

/* Annul the locator to the ORIGIN component. */
            datAnnul( &locor, status );
         }
         if( *status != SAI__OK ) goto L999;

/* Enter the dimensionality and bounds information into the DCB, padding
   the bounds with 1's if necessary. */
         dcb->ndim = ndimd;
         for( i = 0; i < ndimd; i++ ){
            dcb->lbnd[ i ] = orig[ i ];
            dcb->ubnd[ i ] = orig[ i ] + dimd[ i ] - 1;
         }
         for( ; i < ARY__MXDIM; i++ ){
            dcb->lbnd[ i ] = 1;
            dcb->ubnd[ i ] = 1;
         }

/* If the form information in the DCB was not valid, then report an error. */
      } else {
         *status = ARY__FATIN;
         msgSetc( "BADFORM", dcb->form );
         errRep( "ARY1_DBND_FORM",
                 "Unsupported array form '^BADFORM' found in Data Control"
                 "Block (internal programming error).", status );
      }

/* Note if the bounds information is now known. */
L999:
      dcb->kbnd = ( *status == SAI__OK );
   }

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Dbnd", status );

}
